File: resource/src/headless-tester.js
import openSession from "./chrome-devtools-protocol-session.js";
import * as Promises from "./promises.js";

/**
 * Headless 브라우저 테스팅을 위한 래퍼 클래스
 */
class HeadlessTester {
	/**
	 * @constructor
	 * @param {ChromeDevtoolsProtocolSession} chromeDevtoolsProtocolSession
	 * @param {string} hostName
	 */
	constructor(chromeDevtoolsProtocolSession, hostName) {
		this._session = chromeDevtoolsProtocolSession;
		if (hostName.endsWith("/")) {
			hostName = hostName.substring(0, hostName.length - 1);
		}
		if (hostName.includes("://")) {
			var [protocol, host] = hostName.split("://");
			this._hostName = host;
			this._protocol=protocol;
		} else {
			this._hostName = hostName;
			if(hostName.includes("localhost")) {
				this._protocol = "http";
			} else {
				this._protocol="https";
			}
		}
	}

	/**
	 * 리소스 경로를 포함한 전체 URL을 구성합니다.
	 * @private
	 * @param {string} [resourcePath=""]
	 * @returns {string} 완성된 URL
	 */
	_constructUrl (resourcePath = "") {
		if (resourcePath.includes("://")) {
			return resourcePath;
		}
		return this._protocol+"://" + this._hostName+(resourcePath.startsWith("/")? resourcePath:"/"+resourcePath);
	}

	/**
	 * 새로운 HeadlessTester 인스턴스를 생성하는 팩토리 메서드
	 * @param {string} cdpWebSocketUrl - CDP WebSocket URL
	 * @param {string} hostName - 대상 호스트
	 * @returns {Promise<HeadlessTester>}
	 */
	static async create (cdpWebSocketUrl, hostName) {
		var session = await openSession(cdpWebSocketUrl);
		return new HeadlessTester(session, hostName);
	}

	/**
	 * 특정 URL로 페이지를 이동하고 로드가 완료될 때까지 기다립니다.
	 * @param {string} url - 이동할 경로 또는 URL
	 * @returns {Promise<void>}
	 */
	navigate (url) {
		url = this._constructUrl(url);
		this._session.sendMessage("Page.navigate", {url});
		var loadEventPromise = this.waitForEvent(["Page.loadEventFired", "Page.frameStoppedLoading"]);
		var navigatePromise = Promises.unwrappedPromise();
		var unsubscribePromise = this._session.addEventHandler(
			"Network.responseReceived",
			params => {
				var {status} = params.response;
				if(status >= 200 && status < 400) {
					navigatePromise.resolve(params.response);
				} else {
					navigatePromise.reject(new Error(`status[${status}]:::`, {cause:params.response}));
				}
				return false;
			},
			params => {
				var response_url = params.response.url;
				return (url === response_url || url === response_url.substring(0, response_url.length - 1));
			}
		);

		var pp = Promises.timeoutPromise(
			navigatePromise.promise,
			() => {
				unsubscribePromise.then(unsubscribe => unsubscribe());
			}
		);
		return Promise.all([pp, loadEventPromise])
		.then(() => console.log("Page navigated:::"+url));
	}

	/**
	 * 특정 이벤트(들)가 발생할 때까지 대기합니다.
	 * @param {string|string[]} eventNames - 감시할 이벤트 명 또는 배열
	 * @param {number} [timeoutMs] - 제한 시간
	 * @returns {Promise<any[]>} 이벤트 결과 배열
	 */
	waitForEvent(eventNames, timeoutMs) {
		var _events = (Array.isArray(eventNames)? eventNames: [eventNames]);
		var listeners = _events
			.map(eventName => {
				var {resolve, promise} = Promises.unwrappedPromise();
				var unsubscribePromise = this._session.addEventHandler(
					eventName,
					(data) => {
						resolve(data);
						return false;
					}
				);
				return Promises.timeoutPromise(
					promise,
					() => {
						unsubscribePromise.then(unsubscribe =>unsubscribe());
					},
					timeoutMs
				)
			});
		return Promise.all(listeners);
	}

	/**
	 * 브라우저 컨텍스트 내에서 스크립트를 실행합니다.
	 * @param {...(string|Function)} script - 실행할 스크립트 문자열 또는 함수
	 * @returns {Promise<Object>} 실행 결과
	 */
	evaluate(...script) {
		var concat = script.map(s => {
			var str = (typeof s ==="function")?s.toString():s;
			return (str.endsWith(";")) ? str : str + ";";
		}).join("\n");
		return this._session.sendMessage("Runtime.evaluate", {expression:concat});
	}

	/**
	 * 도메인을 활성화/비활성화합니다.
	 * @param {string} domainName
	 * @param {boolean} bool
	 */
	enableDomain (domainName, bool) {
		return this._session.enableDomain(domainName, bool);
	}
}

export {HeadlessTester};

File: language/java/demo/src/main/java/com/example/tester/HeadlessTester.java
package com.example.tester;

import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.example.cdp.CdpDomainEvent;
import com.example.cdp.ChromeDevtoolsProtocolResponse.Result;
import com.example.cdp.PageEvent;

public class HeadlessTester {
    private static final Logger log = LoggerFactory.getLogger(HeadlessTester.class);
    private CdpSession cdpSession;
    private String hostName;
    private String protocol;
    public HeadlessTester(CdpSession cdpSession, String hostName, String protocol) {
        super();
        this.cdpSession = cdpSession;
        if(hostName.endsWith("/")) {
            hostName = hostName.substring(0, hostName.length() -1);
        }
        if(hostName.contains("://")) {
        String[] split = hostName.split("://");
        this.protocol = split[0];
        this.hostName = split[1];
    } else {
        this.hostName = hostName;
        this.protocol = protocol;
    }
    if (this.hostName == null || this.hostName.isEmpty() || this.protocol == null || this.protocol.isEmpty()) {
        throw new IllegalArgumentException("[IllegalArgument] hostName and protocol is required.");
    }
}

private String constructUrl( String resourcePath) {
    if (resourcePath.contains("://")) return resourcePath;
    return protocol + "://" + hostName + (resourcePath.startsWith("/")? resourcePath : "/" + resourcePath);
}

public CompletableFuture<Map<String, Object>> waitForEvent(String eventName) { return waitForEvent(eventName, null);
}

public CompletableFuture<Map<String, Object>> waitForEvent(CdpDomainEvent eventType) {
    if (eventType == null) throw new IllegalArgumentException("eventType cannot be null.");
    return waitForEvent(eventType.eventName(), null);
}
public CompletableFuture<Map<String, Object>> waitForEvent(CdpDomainEvent eventType, Long timeoutMs) {
    if (eventType == null) throw new IllegalArgumentException("eventType cannot be null.");
    return waitForEvent(eventType.eventName(), timeoutMs);
}

public CompletableFuture<Map<String, Object>> waitForEvent (String eventName, Long timeoutMs) {
    long _timeoutMs = (timeoutMs == null ? 5000: timeoutMs);
    CompletableFuture<Map<String, Object>> future = (new CompletableFuture<Map<String, Object>>()).orTimeout(_timeoutMs, TimeUnit.MILLISECONDS);

    Runnable unsubscriber = cdpSession.addEventHandler (eventName, param -> {future.complete(param);
        return false;
    }, null);
    future.exceptionally(err -> {
        log.error("waitForEvent timeout:::{}", eventName);
        // err.printStackTrace();
        unsubscriber.run();
        return null;
    });

    return future;
}

public CompletableFuture<Map<String, Object>> navigate (String url) {
    String _url = constructUrl(url);
    cdpSession.sendMessage("Page.navigate", Map.of("url", _url));
    var future1 = waitForEvent(PageEvent.LOAD_EVENT_FIRED);
    var future2 = waitForEvent(PageEvent.FRAME_STOPPED_LOADING);
    var navigatePromise = new CompletableFuture<Map<String, Object>>();
    Runnable unsubscribe = cdpSession.addEventHandler("Network.responseReceived",
    (Map<String, Object> param) -> {
        Map<String, Object> response = (Map<String, Object>) param.get("response");
        Integer status = (Integer) response.get("status");
        if (status >= 200 && status <400) {
            navigatePromise.complete(response);
        } else {
            navigatePromise.completeExceptionally(new RuntimeException("status["+status+"]:::" + response.toString()));
        }
        return false;
    },
    param -> {
        String responseUrl = (String)(((Map<String, Object>)param.get("response")).get("url"));
        var condition = _url.equals(responseUrl) || _url.equals(responseUrl.substring(0, responseUrl.length() - 1));
        log.debug("condition::::::::::{}\n{}", responseUrl, condition);
        return condition;
    }
);
return CompletableFuture
.allOf(future1, future2, navigatePromise)
.thenCompose(nothing -> CompletableFuture.supplyAsync(() -> navigatePromise.join())
)
.orTimeout(5000, TimeUnit.MILLISECONDS)
.exceptionally(err -> {
    unsubscribe.run();
    throw new RuntimeException(err);
});
}

public CompletableFuture<Result> evaluate(String script, boolean awaitPromise) {
    String _script = "{\n" + script + "\n}";
    return cdpSession.sendMessage("Runtime.evaluate", Map.of("expression", _script, "awaitPromise" ,awaitPromise));
}

public CompletableFuture<Result> evaluate(String script, boolean awaitPromise, long timeoutMs) {
    String _script = "{\n" + script + "\n}";
    return cdpSession.sendMessage("Runtime.evaluate", Map.of("expression", _script, "awaitPromise" ,awaitPromise), timeoutMs);
}

}